From ef59bd764f88d893f1219fee8ba696a5d3f8c1c4 Mon Sep 17 00:00:00 2001
From: bonniegong <46280630+bonniegong@users.noreply.github.com>
Date: Mon, 19 Oct 2020 09:18:07 +0800
Subject: [PATCH] There is a Division by Zero in function OptimizeLayerFrames
 (#2743)

in file MagickCore/layer.c. cur->ticks_per_seconds can be zero
with a crafted input argument *image. This is similar to
CVE-2019-13454.
---
 MagickCore/layer.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/MagickCore/layer.c b/MagickCore/layer.c
index 4ac847a1c1..63bda6d1a7 100644
--- a/MagickCore/layer.c
+++ b/MagickCore/layer.c
@@ -1352,11 +1352,13 @@ static Image *OptimizeLayerFrames(const Image *image,const LayerMethod method,
     if ( disposals[i] == DelDispose ) {
       size_t time = 0;
       while ( disposals[i] == DelDispose ) {
-        time += curr->delay*1000/curr->ticks_per_second;
+        time +=(size_t) (curr->delay*1000*
+          PerceptibleReciprocal((double) curr->ticks_per_second));
         curr=GetNextImageInList(curr);
         i++;
       }
-      time += curr->delay*1000/curr->ticks_per_second;
+      time += (size_t)(curr->delay*1000*
+        PerceptibleReciprocal((double) curr->ticks_per_second));
       prev_image->ticks_per_second = 100L;
       prev_image->delay = time*prev_image->ticks_per_second/1000;
     }
